File: Data Structures & Algorithms/search-for-word-ii/submission-0.py
class TrieNode:
    def __init__(self):
        self.children={}
        self.EndofWord=False
    def addword(self,word):
        cur = self
        for c in word:
            if c not in cur.children:
                cur.children[c]=TrieNode()
            cur = cur.children[c]
        cur.EndofWord = True
class Solution:
    
    

    

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        res = set()
        visit = set()
        root = TrieNode()
        for word in words:
            root.addword(word)
        row = len(board)
        col = len(board[0])
        dir = [[0,1],[0,-1],[1,0],[-1,0]]

        def dfs(r,c,cur,word):
            if r<0 or c<0 or r>=row or c>=col or (r,c) in visit or board[r][c] not in cur.children:
                return
            visit.add((r,c))
            cur = cur.children[board[r][c]]
            word+=board[r][c]
            if cur.EndofWord:
                res.add(word)
            for dr,dc in dir:
                dfs(r+dr,c+dc,cur,word)
            visit.remove((r,c))
        
        for r in range(row):
            for c in range(col):
                dfs(r,c,root,"")
        
        return list(res)


        

File: Data Structures & Algorithms/surrounded-regions/submission-0.cpp
class Solution {
public:
    int dir[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();
        for(int i=0;i<n;i++){
            dfs(i,0,board);
            dfs(i,m-1,board);
        }
        for(int i=0;i<m;i++){
            dfs(0,i,board);
            dfs(n-1,i,board);
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(board[i][j]=='O')board[i][j]='X';
                else if(board[i][j]=='g')board[i][j]='O';
            }
        }
    }
    void dfs(int r, int c, vector<vector<char>>& board){
        if(r>=board.size() || r<0 || c>=board[0].size() || c<0 || board[r][c]!='O')return;
        board[r][c] = 'g';
        for(int i=0;i<4;i++){
            dfs(r+dir[i][0],c+dir[i][1],board);
        }
    }
};
